### PR TITLE
Implement notifications on Review Period update to Awaiting Approval

### PR DESCRIPTION
**Pull request by montesmoci**

### CODE REVIEW

**Review by ZacharyKlein — APPROVED** [review]

File: server/src/main/java/com/objectcomputing/checkins/services/memberprofile/MemberProfileRepository.java [hunk 1]
```diff
@@ -59,6 +60,14 @@ List<MemberProfile> search(@Nullable String firstName, @Nullable String middleNa
                                @Nullable String workEmail, @Nullable String supervisorId, @Nullable Boolean terminated);
 
     List<MemberProfile> findAll();
+    List<UUID> findSupervisoridByIdIn(Set<UUID> Ids);
+
+    @Query(value = "SELECT " +
+            "PGP_SYM_DECRYPT(cast(workEmail as bytea), :aesKey) as workEmail " +
+            "FROM \"member_profile\" mp " +
+            "WHERE  mp.id IN (:ids) ",
+            nativeQuery = true)
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/memberprofile/MemberProfileRepository.java:69]: We can use Java 15 code blocks which I think are easier to read here? WDYT?

```suggestion
    @Query(value = """
            SELECT PGP_SYM_DECRYPT(cast(workEmail as bytea), :aesKey) as workEmail
                    FROM member_profile mp
                    WHERE  mp.id IN (:ids)""",
            nativeQuery = true)
```

File: server/src/main/java/com/objectcomputing/checkins/services/reviews/ReviewPeriodServicesImpl.java [hunk 2]
```diff
@@ -21,14 +32,34 @@ class ReviewPeriodServicesImpl implements ReviewPeriodServices {
...
+                                    @Named(MailJetConfig.HTML_FORMAT) EmailSender emailSender,
+                                    Environment environment, @Property(name = WEB_ADDRESS) String webAddress) {
         this.reviewPeriodRepository = reviewPeriodRepository;
+        this.reviewAssignmentRepository = reviewAssignmentRepository;
+        this.memberProfileRepository = memberProfileRepository;
         this.reviewStatusTransitionValidator = reviewStatusTransitionValidator;
+        this.emailSender = emailSender;
+        this.environment = environment;
+        this.webAddress = webAddress;
+    }
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/reviews/ReviewPeriodServicesImpl.java:59]: This didn't compile on my machine.. shouldn't it be this...?  Maybe the merge messed it up?

```suggestion
    public ReviewPeriodServicesImpl(ReviewPeriodRepository reviewPeriodRepository,
                                    ReviewAssignmentRepository reviewAssignmentRepository,
                                    MemberProfileRepository memberProfileRepository,
                                    ReviewStatusTransitionValidator reviewStatusTransitionValidator,
                                    @Named(MailJetConfig.HTML_FORMAT) EmailSender emailSender,
                                    Environment environment,
                                    @Property(name = WEB_ADDRESS) String webAddress) {
        this.reviewPeriodRepository = reviewPeriodRepository;
        this.reviewAssignmentRepository = reviewAssignmentRepository;
        this.memberProfileRepository = memberProfileRepository;
        this.reviewStatusTransitionValidator = reviewStatusTransitionValidator;
        this.emailSender = emailSender;
        this.environment = environment;
        this.webAddress = webAddress;
    }
```

File: server/src/main/java/com/objectcomputing/checkins/services/reviews/ReviewPeriodServicesImpl.java [hunk 3]
```diff
@@ -21,14 +32,34 @@ class ReviewPeriodServicesImpl implements ReviewPeriodServices {
...
+
+    public void setEmailSender(EmailSender emailSender) {
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/reviews/ReviewPeriodServicesImpl.java:61]: This doesn't need to be public as it's only used in test

```suggestion
   void setEmailSender(EmailSender emailSender) {
```

File: server/src/main/java/com/objectcomputing/checkins/services/reviews/ReviewPeriodServicesImpl.java [hunk 4]
```diff
@@ -96,6 +127,37 @@ public ReviewPeriod update(@NotNull ReviewPeriod reviewPeriod) {
...
+
+        // send notification to supervisors
+        String emailContent = constructEmailContent(reviewPeriodId, reviewPeriodName);
+        emailSender.sendEmail(null, null, "Review Assignments Awaiting Approval", emailContent, supervisorEmails.toArray(new String[0]));
+    }
+
+    private String constructEmailContent (UUID reviewPeriodId, String reviewPeriodName){
+        String emailHtml = "<h3>Review Assignments for Review Period '" + reviewPeriodName + "' are ready for your approval.</h3>";
+        emailHtml += "<a href=\"" + webAddress + "/feedback/reviews?period=" + reviewPeriodId + "\">Click here</a> to review and approve reviewer assignments in the Check-Ins app.";
+        return emailHtml;
```
Comment (timyates) [inline server/src/main/java/com/objectcomputing/checkins/services/reviews/ReviewPeriodServicesImpl.java:160]: Can use text blocks again to remove the requirement to escape double quotes

```suggestion
        return """
                <h3>Review Assignments for Review Period '%s' are ready for your approval.</h3>\
                <a href="%s/feedback/reviews?period=%s">Click here</a> to review and approve reviewer assignments in the Check-Ins app.""".formatted(reviewPeriodName, webAddress, reviewPeriodId);
```

**Review by timyates — CHANGES_REQUESTED** [review]
Couple of suggestions, and I'm not sure if it's the merge, but it's not compiling...

I think the branch needs to be called `feature-xxxxx/some-description` for CI to run it on GitHub 🤔

File: server/src/main/java/com/objectcomputing/checkins/services/reviews/ReviewPeriodServicesImpl.java (hunk 2 above)
Comment (montesmoci) [reply server/src/main/java/com/objectcomputing/checkins/services/reviews/ReviewPeriodServicesImpl.java:59]: Oh I see what I did wrong. I have an unpushed change that fixes the above. Thank you very much for the suggestions and pointing out the branch naming issue! I will include the suggestions then create a new branch/PR.

**Comment by montesmoci** [comment]
Closing in favor of https://github.com/objectcomputing/check-ins/pull/2437

**Pull request closed without merging**